bootstrap.js

```diff
@@ -4,101 +4,206 @@ Cu.import("resource://gre/modules/Console.jsm");
     }
-  },
-  onEvent(e) {},
-};
-
-function insertCopyController() {
-  // Get the "DOM" elements
-  const urlBar = bDocument.getElementById("urlbar");
-  // XUL elements are different than regular children
-  const urlInput = bDocument.getAnonymousElementByAttribute(urlBar, "anonid", "input");
+  }
 
-  // Add the controller to intercept copy "event"
-  // Store controllers so that our controller can inherit cmd_copy from
-  // the actual controller
-  urlInputControllers = urlInput.controllers;
-  urlInputControllers.insertControllerAt(0, urlInputController);
+  onEvent(e) {}
 }
 
-this.install = function(data, reason) {};
+function injectExtension(aWindow) {
+  const browserWindow = new BrowserWindow(aWindow);
+  const copyController = new CopyController(browserWindow);
 
-this.startup = function(data, reason) {
-  bWindow = Services.wm.getMostRecentWindow("navigator:browser");
-  bDocument = bWindow.document;
+  browserWindow.setCopyController(copyController);
+  browserWindowWeakMap.set(aWindow, browserWindow);
 
   // The customizationending event represents exiting the "Customize..." menu from the toolbar.
-  // We need to handle this event because after exiting the customization menu, the copy controller
-  // is removed and we can no longer detect text being copied from the URL bar.
-  // See https://dxr.mozilla.org/mozilla-central/rev/7a9536f89bc75b0672060f16ffbe6eb2c1ff3deb/browser/base/content/browser-customization.js#11
-  bWindow.addEventListener("customizationending", insertCopyController);
+  // We need to handle this event because after exiting the customization menu, the copy
+  // controller is removed and we can no longer detect text being copied from the URL bar.
+  // See DXR:browser/base/content/browser-customization.js
+  browserWindow.addCustomizeListener();
 
   // Load the CSS with the shareButton animation
-  utils = bWindow.QueryInterface(Ci.nsIInterfaceRequestor).getInterface(Ci.nsIDOMWindowUtils);
-  utils.loadSheet(CSS_URI, utils.AGENT_SHEET);
+  browserWindow.insertCSS();
 
-  insertCopyController();
+  // insert the copy controller to detect copying from URL bar
+  browserWindow.insertCopyController();
+}
+
+// see https://dxr.mozilla.org/mozilla-central/rev/53477d584130945864c4491632f88da437353356/xpfe/appshell/nsIWindowMediatorListener.idl
+const windowListener = {
+  onWindowTitleChange(window, title) { },
+  onOpenWindow(xulWindow) {
+    // xulWindow is of type nsIXULWindow, we want an nsIDOMWindow
+    // see https://dxr.mozilla.org/mozilla-central/rev/53477d584130945864c4491632f88da437353356/browser/base/content/test/general/browser_fullscreen-window-open.js#316
+    // for how to change XUL into DOM
+    const domWindow = xulWindow.QueryInterface(Ci.nsIInterfaceRequestor)
+      .getInterface(Ci.nsIDOMWindow);
+
+    // we need to use a listener function so that it's injected
+    // once the window is loaded / ready
+    const onWindowOpen = (e) => {
+      domWindow.removeEventListener("load", this);
```

Comment:
This is a good pattern, but I don't think `this` is the `onWindowOpen` function. You should be removing `onWindowOpen` not `this`.